### PR TITLE
Decouple Performance Metrics from Execution Strategy via Template Method Pattern

### PR DESCRIPTION
**Pull request by Nia-bald**
### Summary

Introduces an `ExecutionStrategy` abstract base class that separates **what to time** from **how to execute**, ensuring consistent metric definitions across all current and future execution strategies. Also adds a comprehensive project README.

### Problem

Performance metrics (prefill time, decode throughput, etc.) were calculated **inside** `PipelineEngine::generate()`. This created three problems:

1. **Timing was strategy-owned** — `PipelineEngine` decided what to time and how, meaning a future strategy (e.g., KV cache) would need to reimplement all timing logic.
2. **Metric formulas were embedded in execution code** — tokens/sec calculations were hardcoded alongside CUDA kernels, coupling concerns that should be independent.
3. **No guarantee of semantic consistency** — if two strategies both report "prefill_time_ms", nothing enforced that they measured the same thing.

### Solution

**Template Method pattern** — the base class `ExecutionStrategy` owns `generate()` and defines the execution skeleton:

```
generate() {
    prefill_time = time( run_prefill() )   ← base class defines & times
    decode_time  = time( run_decode()  )   ← base class defines & times
    compute tok/s, total_time              ← base class computes once
    finalize()                             ← strategy detokenizes
}
```

Strategies only implement `run_prefill()`, `run_decode()`, and `finalize()` — they never touch a timer or compute a metric. This **structurally guarantees** that "prefill time" means the same thing for every strategy.

### Changes

| File | Change |
|---|---|
| `include/pipeline/execution_strategy.hpp` | **NEW** — Abstract base class with Template Method. Owns `generate()`, `time_cuda_execution()`, and all metric computation. |
| `include/pipeline/pipeline_engine.hpp` | Inherits from `ExecutionStrategy`. Exposes `run_prefill()`, `run_decode()`, `finalize()` as protected overrides. Removed `generate()` and `time_cuda_execution()`. |
| `src/pipeline/pipeline_engine.cpp` | Split monolithic `generate()` into three focused hooks with pure execution logic — no timing, no metric math. |
| `include/batch_executor.hpp` | Strategy member changed from `unique_ptr<PipelineEngine>` to `unique_ptr<ExecutionStrategy>`. |
| `README.md` | **NEW** — Comprehensive project documentation. |

### What Did NOT Change (by design)

- `bench_performance.cu` — zero changes, reads `result.metrics` as before
- `main.cpp` — zero changes
- `batch_executor.cpp` — zero changes (polymorphism handles it)
- `batch_executor_orchestrator.*` — zero changes
- `metrics.hpp` — zero changes

### How to Add a New Strategy After This PR

```cpp
class KVCacheEngine : public pipeline::ExecutionStrategy {
protected:
    void run_prefill(...) override { /* your prefill logic */ }
    void run_decode(...) override  { /* your decode logic */ }
    void finalize(...) override    { /* detokenize */ }
    cudaStream_t get_stream() const override { return stream; }
};
```

Then add one line to `batch_executor.cpp`:
```cpp
case StrategyType::KV_CACHE:
    strategy = std::make_unique<pipeline::KVCacheEngine>(...);
```

Timing, metrics, and benchmarks work automatically — no other changes needed.

### Testing

- ✅ Full build passes (`cmake + make -j`)
- ✅ `gpt2_engine` produces correct output with accurate metrics
- ✅ `bench_performance` runs end-to-end

### CODE REVIEW

File: src/pipeline/pipeline_engine.cpp [hunk 1]
```diff
@@ -18,106 +17,65 @@ PipelineEngine::PipelineEngine(Transformer& model, GPT2Tokenizer& tokenizer, GPU
...
         int current_seq_len = result.output_sequence.size();
+
+        // Re-use scratch memory for each token
         inference_arena.reset_to(persistent_offset);
-        
+
         cudaMemcpyAsync(d_input_ids, result.output_sequence.data(), current_seq_len * sizeof(int), cudaMemcpyHostToDevice, stream);
-        
+
         model.forward(d_input_ids, d_logits, 1, current_seq_len, inference_arena, stream);
```
Comment (Nia-bald) [inline src/pipeline/pipeline_engine.cpp:61]: batch size needs to be passed through configs not hardcoded here
